scipy/sparse/csgraph/_flow.pyx

```diff
@@ -251,6 +253,8 @@ def maximum_flow(csgraph, source, sink, *, method='dinic'):
     if not csgraph.has_sorted_indices:
         csgraph = csgraph.sorted_indices()
 
+    csgraph.indices, csgraph.indptr = _safe_downcast_indices(csgraph)
```

Comment:
Does this mean that the input sparray passed to the function will sometimes have its indices modified by the maximum_flow() call? If so, I'd prefer that we not assign back into `csgraph` and instead change the `_add_reverse_edges()` function to take the member arrays directly.

Reply:
I updated the code to check whether the index arrays had chaged and if so to construct a new csgraph without copying data. It turns out that `_add_reversed_edges()` computes `csr_array(a.transpose())` which, if I passed in `csgraph` was looking at the original `idx_dtype`. I also tried moving the transpose outside the function and passing all the csgraph attributes. But this seems cleaner and doesn't copy unless something needs to change. 

I went through the rest of the PR and made sure the input csgraphs were not getting changed in the other places either.